### PR TITLE
Idr0079 scripting example

### PR DESCRIPTION
**Pull request by will-moore**
This adds the idr0079 heatmap scripting example to the figure scripting page and makes a few other improvements to the page:

 - Moves the generic intro to JavaScript scripting out of the FRAP example
 - Fixes the fixed height on screenshots (which caused them to be squashed out of aspect ratio when the page was resized to be thin)
 - Adds the scripting example from idr0079.

To test:
 - Check the page reads OK
 - Test the code using figure at https://workshop.openmicroscopy.org/figure/file/136949

### CODE REVIEW

**Comment by pwalczysko** [comment]
For the new example (Shapes heatmap from OMERO table), you would need some specific data. I think these should be specified in the Resources (I guess you need some ROIs and analysis results on those ROIs ?)

File: docs/omero_figure_scripting.rst [hunk 1]
```diff
@@ -109,9 +121,30 @@ We will use the time-lapse images listed above to create a FRAP figure but you c
...
    :height: 0.27231in
+
+
+Example 2: Shapes heatmap from OMERO.table
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example uses an OMERO.table linked to each Image to generate
+a heatmap of colors applied to Shapes on the figure panel.
+Selected panels need to have Shapes added from OMERO (in the ROIs dialog).
+This means that each shape JSON will have an ``id`` that corresponds to a Shape in
```
Comment (pwalczysko) [inline docs/omero_figure_scripting.rst]:
```suggestion
Thus, each shape JSON will have an ``id`` that corresponds to a Shape in
```

Maybe like this ? This is, I am assuming that the ids in JSON will be populated by virtue of adding the ROIs from OMERO ? If not, how does the JSON get the shapes ids ?

File: docs/omero_figure_scripting.rst [hunk 2]
```diff
@@ -109,9 +121,30 @@ We will use the time-lapse images listed above to create a FRAP figure but you c
...
+OMERO.
+
+The code at `figure_table_data_shapes.js <https://github.com/ome/omero-guide-figure/tree/master/scripts/figure_table_data_shapes.js>`_
```
Comment (pwalczysko) [inline docs/omero_figure_scripting.rst]: This link waits for merging of the script PR ? (does not work atm)

File: docs/omero_figure_scripting.rst [hunk 3]
```diff
@@ -63,16 +59,32 @@ We will use the time-lapse images listed above to create a FRAP figure but you c
...
 
-#.  The ``figureModel`` variable is accessible in the console. We can use ``figureModel.getSelected()`` to get selected panels and for each panel we can call ``p.set()`` to change an attribute.
+#.  The ``figureModel`` variable is accessible in the `Console` of the browser `Developer Tools`. The easiest way to open
+    the developer tools in most browsers is to open the context menu (right-click) anywhere on the page and choose `Inspect`.
+
+#.  We can use ``figureModel.getSelected()`` to get selected panels and for each panel we can call ``p.set()`` to change an attribute.
 
-#.  For example, to set the ``height`` of each selected panel to ``200``, we can do:
+#.  For example, to set the ``height`` of several panels to ``200``, we can select the panels in the figure UI and paste
+    this code snippet in the `Console`:
```
Comment (pwalczysko) [inline docs/omero_figure_scripting.rst:72]: For the code snippet below, when the workflow was followed as written, I get just

```
Location: https://workshop.openmicroscopy.org/figure/load_web_figure/136949/?_=1622452009145
Line Number 1, Column 1: 136949:1:1
Empty string passed to getElementById().
```

Comment (pwalczysko) [reply docs/omero_figure_scripting.rst:72]: This code snippet works in Chrome. But in my FF, I am getting the problem described in https://github.com/ome/omero-guide-figure/pull/11/files#r642426364 - I guess this functionality is Chrome only or some steps are missing in description.

**Comment by pwalczysko** [comment]
In summary, I think 3 things are needed:

1. offer the OMERO.table which is sitting at the back of the new ROI colorcoding workflow. How to get it ? What is necessary to create it, or better, an example directly attached to this walkthrough.
2. Clarify the Firefox problem/workflow, atm the whole section seems to me to be ``Chrome only``
3. In the testing, some positive things were happening, but what are the parameters other than "Sphericity" I could try ?

File: docs/omero_figure_scripting.rst (hunk 2 above)
Comment (will-moore) [reply docs/omero_figure_scripting.rst]: That like waits for merging of *this* PR. `figure_table_data_shapes.js` is added in this PR.

**Comment by will-moore** [comment]
Thanks @pwalczysko 

1. Providing a generic solution for creating a suitable OMERO.table is a bit tricky just now. I started looking at how to use `populate metadata` to create one from a CSV. This needs https://github.com/ome/omero-scripts/pull/184 to be able to run the populate metadata script on an Image from webclient. But you also need to have ROIs with the `name` attribute set on them, so you can't just do it through the UI. I'm going to try and fix that in the metadata plugin.
2. It worked fine for me in Firefox. I also see the warning you saw (before even pasting the code sample) and I think Firefox might only allow pasting if you type in "allow pasting" or something like that.
3. If you look at the Image in that figure, you'll see it has this OMERO.table on it: https://workshop.openmicroscopy.org/webclient/omero_table/311957/ so you can choose any numerical column name.

**Comment by pwalczysko** [comment]
> Thanks @pwalczysko
> 
>     1. Providing a generic solution for creating a suitable OMERO.table is a bit tricky just now. I started looking at how to use `populate metadata` to create one from a CSV. This needs [ome/omero-scripts#184](https://github.com/ome/omero-scripts/pull/184) to be able to run the populate metadata script on an Image from webclient. But you also need to have ROIs with the `name` attribute set on them, so you can't just do it through the UI. I'm going to try and fix that in the metadata plugin. 
> 
>     2. It worked fine for me in Firefox. I also see the warning you saw (before even pasting the code sample) and I think Firefox might only allow pasting if you type in "allow pasting" or something like that.
> 
>     3. If you look at the Image in that figure, you'll see it has this OMERO.table on it: https://workshop.openmicroscopy.org/webclient/omero_table/311957/ so you can choose any numerical column name.

Ad 1.  Thank you for the effort to make the workflow CSV -> OMERO.table possible. It might take time to go through the two PRs you are pointing to. In the meantime, I suppose that you created the OMERO.table using a CLI ? If so, it could be possible to attach the CSV here now and say "create a table from it following https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata.html ?  The workflow is not really targeted at a UI user, so even sending the reader to the CLI is preferable to keeping things unclear imho.
Ad 2.  Indeed, I typed into the Console ``allow pasting`` and then the snippet was pasted and worked too. The confusing bit is that FF is still giving a red error? messages even as a reaction to the ``allow pasting``. Not sure if we want to metion the "allow pasting" trick here ?
Ad 3. Yes, I personally can do what you suggest, but the reader of this text will probably not be able to, that is what I meant. Maybe that could be solved by attaching the CSV as discussed in Ad 1. and remarking on this in this step in the scenario would be an improvement I would suggest.

**Comment by will-moore** [comment]
@pwalczysko Just read your comment after pushing that last commit, but I think it addresses some points.

Actually, I realised I already have a PR to describe the idr0079 workshop setup: https://github.com/ome/training-scripts/pull/88 but that might be a slight overkill if you want to use this script on your own images. In the commit just now I refer users to the omero-metadata docs.

File: docs/omero_figure_scripting.rst [hunk 4]
```diff
@@ -13,44 +13,43 @@ Setup
 
 -  Install the OMERO.figure web app as described at https://pypi.org/project/omero-figure/
 -  Upload the script :download:`Split_View_Figure.py <../scripts/Split_View_Figure.py>` to the OMERO scripting service
+-  For the `Shapes heatmap from OMERO.table` example, you'll need to draw ROIs on an Image and create an OMERO.table on the
+   Image with a ``Shape`` column with corresponding Shape IDs, as described at `omero-metadata <https://github.com/ome/omero-metadata#populate>`_
```
Comment (pwalczysko) [inline docs/omero_figure_scripting.rst]: I do not see. any mention of ``Shape`` in https://github.com/ome/omero-metadata#populate - nor is there such in any open PR.  I think we should add it there. But even better would be an addition of an example CSV directly to this workflow. We have a precedent of this in https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata.html, I mean the https://omero-guides.readthedocs.io/en/latest/_downloads/8227ce5769b3477eff58853580d44493/four-images.csv - for that, the CSV does not need to be the one used in idr0079 workflow, but a simple one.

File: docs/omero_figure_scripting.rst [hunk 5]
```diff
@@ -109,9 +124,36 @@ We will use the time-lapse images listed above to create a FRAP figure but you c
...
+To see the available columns for the table, find the table under the Image's `Annotations` tab in the
+webclient and click the `eye` icon to view the table.
+
+Once the values for all Shapes on the panel are loaded, the code calculates the range and
+generates a heatmap color for each value in that range. This is set as the color
+on each Shape.
+
+In the screenshot below, Shapes in the first panel are colored according to the ``Centroids_RAW_X``
+column and Shapes on the lower panel are colored according to the ``Sphericity`` column.
+Images are from `idr0079`.
```
Comment (pwalczysko) [inline docs/omero_figure_scripting.rst]:
```suggestion
Images in this example are from `idr0079`.
```

File: docs/omero_figure_scripting.rst [hunk 6]
```diff
@@ -13,44 +13,43 @@ Setup
 
 -  Install the OMERO.figure web app as described at https://pypi.org/project/omero-figure/
 -  Upload the script :download:`Split_View_Figure.py <../scripts/Split_View_Figure.py>` to the OMERO scripting service
+-  For the `Shapes heatmap from OMERO.table` example, you'll need to draw ROIs on an Image and create an OMERO.table on the
```
Comment (pwalczysko) [inline docs/omero_figure_scripting.rst]: A link to the subchapter could be added similarly to https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata-ui.html#populatemetadatapdi (you will need an anchor just in front of the subchapter, see
Screenshot 2021-06-01 at 14 43 50


 in https://raw.githubusercontent.com/ome/omero-guide-upload/master/docs/metadata-ui.rst

File: docs/omero_figure_scripting.rst (hunk 4 above)
Comment (will-moore) [reply docs/omero_figure_scripting.rst]: On that page the `four-images.csv` is linked but isn't mentioned at all. Do users need to edit it?
I'll add an example csv and also link to the training-scripts idr0079 prep doc for full set-up

**Comment by will-moore** [comment]
Added internal links and expanded the setup section for OMERO.tables.

File: docs/omero_figure_scripting.rst (hunk 4 above)
Comment (pwalczysko) [reply docs/omero_figure_scripting.rst]: syrry, my bad. The four-images.csv is used in https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata-ui.html - the page I sent you to has the link to it for historical reasons only

**Comment by pwalczysko** [comment]
Do you want to list it for tomorrow please ?

**Comment by pwalczysko** [comment]
In the Description of the scripting subchapter, it says 
Screenshot 2021-06-02 at 17 37 23
this is outdated, as now there are 2 examples, and only one of them is doing what is described above.

File: docs/omero_figure_scripting.rst [hunk 7]
```diff
@@ -13,44 +13,45 @@ Setup
 
 -  Install the OMERO.figure web app as described at https://pypi.org/project/omero-figure/
 -  Upload the script :download:`Split_View_Figure.py <../scripts/Split_View_Figure.py>` to the OMERO scripting service
+-  For the :ref:`Shapes heatmap from OMERO.table<omeroTableData>` example, you'll need to draw ROIs on an Image and create an OMERO.table on the
+   Image with a ``Shape`` column with corresponding Shape IDs, as described below.
```
Comment (pwalczysko) [inline docs/omero_figure_scripting.rst]: I would remove ``below`` - it just confused me, as there is a link in the first part of the sentence, but the ``below`` is not a link, so was scrolling through the doc to find "below". Simplest would be to say "inside the example", but better would be to add the same link you have there already to the word "below"

**Comment by pwalczysko** [comment]
Just two minor issues to reformulate. Otherwise the text seems clear to me now.

File: docs/omero_figure_scripting.rst (hunk 7 above)
Comment (will-moore) [reply docs/omero_figure_scripting.rst]: I'd prefer not to add 2 links to the same place with different text as it looks like they're both different (and you might need to click both). So I'll say "in the example".

**Comment by pwalczysko** [comment]
Nevertheless, please see https://github.com/ome/omero-metadata/pull/58#issuecomment-853239006 - the simple workflow did not work for me

**Comment by will-moore** [comment]
That last commit fixes an issue with running the script when the server has a URL prefix, as described at https://github.com/ome/omero-metadata/pull/58#issuecomment-853239006

It should work on `merge-ci` (with prefix), e.g. using figure https://merge-ci.openmicroscopy.org/web/figure/file/76582/ and `probability` column AND on workshop (no prefix) with https://workshop.openmicroscopy.org/figure/file/136949 and e.g. `Sphericity` column name.

**Comment by pwalczysko** [comment]
Functionality works now (the script produced the figure color coding on a completely new image with a completely new OMERO.table attached). Also the text changes build and read okay.

Ready to merge fmpov.

**Review by pwalczysko — APPROVED** [review]

**Comment by jburel** [comment]
Thanks

**Pull request merged**